packages/thirdweb/src/react/web/ui/prebuilt/thirdweb/BuyDirectListingButton/index.tsx

```diff
@@ -0,0 +1,189 @@
+   */
+  client: ThirdwebClient;
+
+  /**
+   * ID of the marketplace's DirectListing
+   */
+  listingId: bigint;
+
+  /**
+   * Qty to buy (optional)
+   *
+   * - For ERC721 listing: the `quantity` is always hard-coded to 1n - passing this props doesn't do anything
+   *
+   * - For ERC1155 listing: the `quantity` defaults to the quantity of the listing if not specified.
+   *
+   * The component will also throw an error if  you pass a `quantity` and it's greater than the listing's quantity
+   */
+  quantity?: bigint;
+};
+
+/**
+ * This button is used with thirdweb Marketplace v3 contract, for buying NFT(s) from a listing.
+ *
+ * Under the hood, it prepares a transaction using the [`buyFromListing` extension](https://portal.thirdweb.com/references/typescript/v5/marketplace/buyFromListing)
+ * and then pass it to a <TransactionButton />
+ *
+ * Since it uses the TransactionButton, it can take in any props that can be passed
+ * to the [`TransactionButton`](https://portal.thirdweb.com/references/typescript/v5/TransactionButton)
+ *
+ *
+ * @param props
+ * @returns <TransactionButton />
+ *
+ * @example
+ * ```tsx
+ * import { BuyDirectListingButton } from "thirdweb/react";
+ *
+ * <BuyDirectListingButton
+ *   contractAddress="0x..." // contract address of the marketplace v3
+ *   chain={...} // the chain which the marketplace contract is deployed on
+ *   client={...} // thirdweb client
+ *   listingId={100n} // the listingId or the item you want to buy
+ *   quantity={1n} // optional - see the docs to learn more
+ * >
+ *   Buy NFT
+ * </BuyDirectListingButton>
+ * ```
+ */
+export function BuyDirectListingButton(props: BuyDirectListingButtonProps) {
+  const {
+    contractAddress,
+    listingId,
+    children,
+    chain,
+    client,
+    quantity,
+    payModal,
+  } = props;
+  const defaultPayModalMetadata = payModal ? payModal.metadata : undefined;
+  const account = useActiveAccount();
+  const contract = getContract({
+    address: contractAddress,
+    client,
+    chain,
+  });
+
+  const { data: payMetadata } = useReadContract(getPayMetadata, {
+    contract,
+    listingId,
+    queryOptions: {
+      enabled: !defaultPayModalMetadata,
+    },
+  });
+
+  const { mutateAsync } = useSendAndConfirmTransaction();
+
+  async function prepareBuyTransaction() {
+    if (!account) {
+      throw new Error("No account detected");
+    }
+    const [listing, { getApprovalForTransaction }, { buyFromListing }] =
+      await Promise.all([
+        getListing({
+          contract,
+          listingId,
+        }),
+        import(
+          "../../../../../../extensions/erc20/write/getApprovalForTransaction.js"
+        ),
+        import(
+          "../../../../../../extensions/marketplace/direct-listings/write/buyFromListing.js"
+        ),
+      ]);
+    if (!listing) {
+      throw new Error(`Could not retrieve listing with ID: ${listingId}`);
+    }
+
+    // For ERC721 listing, always hard-code to 1n
+    const _quantity =
```

Comment:
Instead of implicitly overriding the inputted quantity here we should throw an error if they try specifying a quantity > 1 for ERC721. Or adjust the types to not allow it. Or both, ideally both to cover both JS and TS

Reply:
wow that's good thinking ser. I will add "cover both js and ts" to my list!

Reply:
although I don't think we can adjust the type? Because we won't know if it's ERC721 until we fetch the listing. So maybe just throw error?

Reply:
@gregfromstl actually do you think it would be better UX + DX to handle the ERC721 case for user? Because in the code we only know if it's ERC721 once we have fetched the listing, and if users intend to buy, it's always gonna be at least "1" anyway